### PR TITLE
Feat: User Login API issue#47

### LINKED ISSUE #47: Development: Create Login API
Description
As a student, I need to have the ability to login to BIT. For this the API on the backend should be available.
Mocks
NA

Acceptance Criteria
Create an API for user login for the BIT backend. The API would take a username/password combination as input and return a response code based on the validity of the data in the database. Majorly, 2 outcomes are possible: Successful (200 OK), or Unsuccessful (40x) (x represents one of the error codes like 400, 403, etc.)

Alternatives for testing the API: CuRL/Python requests

Update [Required]
Flask POST request for user login
 Takes user attributes as a json (eventually this will be sent through the UI)
 Response handling: 200 (Successful Process), 401 (unauthorized), 403 (forbidden), 400 (bad request), 500 (internal server error)
 Test the API using Postman (or any alternative)
Enhancement to Update [Optional]
Definition of Done
 All of the required items are completed.
 Approval by 2 mentors.

### PR DESCRIPTION
**Pull request by mtreacy002**
### Description
Added Login User API

Fixes #47 

### Type of Change:

- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Try to login a user
Gif file: https://media.giphy.com/media/L329CK5dENhpM68PVN/giphy.gif

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers . I've just created Login API. Can you please review this PR. @ramitsawhney27 , do you want the test cases to be done on a separate PR or just in this one? I'm thinking of waiting for the Register API test cases PR to be approved and add Login API on top of that code base, so will do a separate PR from this one. What do you think?

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers . I've just pushed latest code change to fix message display connection error response.

File: app/api/models/user.py
Comment (isabelcosta) [inline app/api/models/user.py]: As we talked before, you don't need to thing about Refresh API for now, so these 2 attributes can be removed: `refresh_token` and `refresh_expiry`

File: app/api/models/user.py
Comment (isabelcosta) [inline app/api/models/user.py]: You could remove this as well

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@mtreacy002 I added comments according to our 1:1 today, where we discussed you can remove the Refresh API related code, as this is not a priority for MVP.

File: app/api/models/user.py
Comment (mtreacy002) [reply app/api/models/user.py]: Got it. Will remove it now.

File: app/api/models/user.py
Comment (mtreacy002) [reply app/api/models/user.py]: done

File: app/api/ms_api_utils.py
Comment (ramitsawhney27) [inline app/api/ms_api_utils.py]: Same as my previous comment on catching exceptions

File: app/api/resources/users.py
Comment (ramitsawhney27) [inline app/api/resources/users.py:80]: This method is too big and does too many things. Break this down into multiple smaller functions 
https://martinfowler.com/bliki/FunctionLength.html

**Review by ramitsawhney27 — CHANGES_REQUESTED** [review]

File: app/api/resources/users.py
Comment (mtreacy002) [reply app/api/resources/users.py:80]: I'm not sure I'm following. Isn't this just to cover POST /login and its responses? If you want to make it into smaller functions, can you please give me example of endpoints that will come as the result of slicing them?

Comment (mtreacy002) [reply app/api/resources/users.py:80]: Done. Thanks for giving this idea, @ramitsawhney27 . I've refactored HTTP response checking for both Register and Login. Hopefully this will help tidying up responses for any future API endpoints too.

Comment (mtreacy002) [reply app/api/resources/users.py:80]: Note that for `all other errors` comment under http_response_checker function, that is only for errors that are unplanned. As we move through other functionalities, I'll keep adding smaller functions to check HTTP response type/message.

File: app/api/ms_api_utils.py
Comment (ramitsawhney27) [inline app/api/ms_api_utils.py]: Excess spaces between the 2 functions

File: app/api/ms_api_utils.py
Comment (ramitsawhney27) [inline app/api/ms_api_utils.py:51]: Can make these 3 lines a separate function

File: app/api/ms_api_utils.py
Comment (ramitsawhney27) [inline app/api/ms_api_utils.py]: Shouldn't this be finally?

File: app/api/ms_api_utils.py
Comment (ramitsawhney27) [reply app/api/ms_api_utils.py]: We need to fix this beyond just logging it. If it's a specific message it won't be checked.
And this should not be a logging.warning, it should be fatal

**Review by ramitsawhney27 — CHANGES_REQUESTED** [review]

File: app/api/ms_api_utils.py
Comment (mtreacy002) [reply app/api/ms_api_utils.py:51]: When the PR647 in MS got approved (which will be soon, hopefully), this code will no longer be 3 lines. It'll just be 2 lines like the rest. Can we keep it here for now?

File: app/api/ms_api_utils.py
Comment (mtreacy002) [reply app/api/ms_api_utils.py]: Changed it to logging.fatal now. 👍

File: app/api/ms_api_utils.py
Comment (mtreacy002) [reply app/api/ms_api_utils.py]: got it.

File: app/api/ms_api_utils.py
Comment (mtreacy002) [reply app/api/ms_api_utils.py]: fixed it now.

File: app/api/ms_api_utils.py
Comment (mtreacy002) [reply app/api/ms_api_utils.py]: @ramitsawhney27 Any tips what else need to be done here?

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers. I have pull rebased the changes made per mentor feedback on PR #59 (Register API tests) onto this PR as well so the changes made here already in sync with changes done on PR #59.

File: app/api/ms_api_utils.py
Comment (ramitsawhney27) [reply app/api/ms_api_utils.py]: Resolved for now

File: app/api/ms_api_utils.py
Comment (ramitsawhney27) [reply app/api/ms_api_utils.py:51]: Sure

File: app/api/ms_api_utils.py
Comment (ramitsawhney27) [inline app/api/ms_api_utils.py:64]: Won't this change now? should be indexed using the named tuple?

Comment (mtreacy002) [reply app/api/ms_api_utils.py:64]: Yes, It is now (on PR #65), already using namedtuple. This Login PR #60 was made before that namedtuple PR. I don't think I need to change in this PR since it'll be overwritten by PR #65 changes if that one also got approved after this.

screenshot from namedtuple ms_api_utils.py

Screen Shot 2020-06-27 at 7 35 48 pm

The commit history showing sequence of changes

Screen Shot 2020-06-27 at 7 36 02 pm

**Comment by mtreacy002** [comment]
@anitab-org/bridgeintech-maintainers . Any progress on this PR review? If there's no other changes required, can you please approve this PR? 
Note: This PR is the first in line to be merged out of all the current open PRs. Thanks.

**Review by foongminwong — APPROVED** [review]
Tested User Login API [here](https://github.com/anitab-org/bridge-in-tech-web/pull/26) with no errors

**Comment by mtreacy002** [comment]
@ramitsawhney27 as you have reviewed tthis PR and verbally approves it in BIT weekly meeting last week, can you please give the formal approval so it can be merged? Thanks

**Review by ramitsawhney27 — APPROVED** [review]

**Review by isabelcosta — APPROVED** [review]

**Pull request merged**
